sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,356 @@
+        Speeds of joint.
+    parent_joint_pos : Vector, optional
+        Defines the joint's point where the parent will be connected to child.
+        Default value is masscenter of Parent Body.
+    child_joint_pos : Vector, optional
+        Defines the joint's point where the child will be connected to parent.
+        Default value is masscenter of Child Body.
+    parent_axis : Vector, optional
+        Axis of parent frame which would be aligned with child's
+        axis. Default is x axis in parent's frame.
+    child_axis : Vector, optional
+        Axis of child frame which would be aligned with parent's
+        axis. Default is x axis in child's frame.
+
+    """
+
+    def __init__(self, name, parent, child, coordinates=None, speeds=None, parent_joint_pos=None,
+        child_joint_pos=None, parent_axis=None, child_axis=None):
+
+        if not isinstance(name, str):
+            raise TypeError('Supply a valid name.')
+        self._name = name
+
+        if not isinstance(parent, Body):
+            raise TypeError('Parent must be an instance of Body.')
+        self._parent = parent
+
+        if not isinstance(child, Body):
+            raise TypeError('Child must be an instance of Body.')
+        self._child = child
+
+        self._coordinates = self._generate_coordinates(coordinates)
+        self._speeds = self._generate_speeds(speeds)
+        self._kdes = self._generate_kdes()
+
+        self._parent_axis = self._axis(parent, parent_axis)
+        self._child_axis = self._axis(child, child_axis)
+
+        self._parent_joint = self._locate_joint_pos(parent,parent_joint_pos)
+        self._child_joint = self._locate_joint_pos(child, child_joint_pos)
+
+        self._orient_frames()
+        self._set_angular_velocity()
+        self._set_linear_velocity()
+
+    def __str__(self):
+        return self._name
+
+    def __repr__(self):
+        return self.__str__()
+
+    def parent(self):
+        """ Parent body of Joint."""
+        return self._parent
+
+    def child(self):
+        """ Child body of Joint."""
+        return self._child
+
+    def coordinates(self):
+        """ List of coordinates of Joint."""
+        return self._coordinates
+
+    def speeds(self):
+        """ List of speeds of Joint."""
+        return self._speeds
+
+    def kdes(self):
+        """ KDE of Joint."""
+        return self._kdes
+
+    @abstractmethod
+    def _generate_coordinates(self, coordinates):
+        """ Generate list of coordinates."""
+        pass
+
+    @abstractmethod
+    def _generate_speeds(self, speeds):
+        """ Generate list of speeds. """
+        pass
+
+    @abstractmethod
+    def _orient_frames(self):
+        """Orient frames as per the joint"""
+        pass
+
+    @abstractmethod
+    def _set_angular_velocity(self):
+        pass
+
+    @abstractmethod
+    def _set_linear_velocity(self):
+        pass
+
+    def _generate_kdes(self):
```

Comment:
This should be an abstract method.

Reply:
I don't think it needs to be an abstract method.

Reply:
For more complicated joints, it isn't always the case that you have to use `q' = u`.

Reply:
I'm not familiar with the `abstractmethod` decorator. I see here https://docs.python.org/3/library/abc.html#abc.abstractmethod that you use if if you want to prevent instantiation of the object if those methods aren't overwritten in the sub class, so this doesn't need to be decorated in that case, because it is optionally overridden, as are any methods.